File: packages/domain/src/domain/operations.py
from pydantic import BaseModel, Field


class PipelineHealthSnapshot(BaseModel):
    extraction_pending: int = 0
    extraction_failed: int = 0
    knowledge_pending: int = 0
    triage_pending: int = 0
    embedding_model_mismatch_versions: int = 0


class MaintenanceActionResult(BaseModel):
    pipeline_recovery_enqueued: bool = False
    embedding_mismatch_flagged: int = 0


class OperationsStatus(BaseModel):
    pending_outbox_events: int = 0
    failed_outbox_events: int = 0
    processed_outbox_events: int = 0
    canonical_entities: int = 0
    canonical_claims: int = 0
    projection_rebuild_recommended: bool = False
    pipeline: PipelineHealthSnapshot = Field(default_factory=PipelineHealthSnapshot)
    maintenance_recommended: bool = False
    status_summary_hu: str = ""


class ProjectionRebuildResult(BaseModel):
    entities_projected: int = 0
    relationships_projected: int = 0
    claims_projected: int = 0
    contradictions_projected: int = 0
    cleared_nodes: bool = False

File: apps/web/src/lib/operations.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

export type OperationsStatus = {
  pending_outbox_events: number;
  failed_outbox_events: number;
  processed_outbox_events: number;
  canonical_entities: number;
  canonical_claims: number;
  projection_rebuild_recommended: boolean;
  pipeline: {
    extraction_pending: number;
    extraction_failed: number;
    knowledge_pending: number;
    triage_pending: number;
    embedding_model_mismatch_versions: number;
  };
  maintenance_recommended: boolean;
  status_summary_hu: string;
};

export type MaintenanceRunResult = {
  pipeline_recovery_enqueued: boolean;
  embedding_mismatch_flagged: number;
  status_summary_hu: string;
};

export async function fetchOperationsStatus(): Promise<OperationsStatus | null> {
  const response = await fetch(`${API_URL}/api/v1/operations/status`, { cache: "no-store" });
  if (!response.ok) return null;
  return response.json();
}

export async function runMaintenanceRecovery(): Promise<MaintenanceRunResult | null> {
  const response = await fetch(`${API_URL}/api/v1/operations/maintenance/run`, {
    method: "POST",
  });
  if (!response.ok) return null;
  return response.json();
}

export async function rebuildProjection(asyncMode = true): Promise<boolean> {
  const response = await fetch(
    `${API_URL}/api/v1/operations/projection/rebuild?async=${asyncMode ? "true" : "false"}`,
    { method: "POST" },
  );
  return response.ok;
}
